substrate/client/transaction-pool/tests/fatp.rs

```diff
@@ -1975,6 +1975,118 @@ fn fatp_future_is_pruned_by_conflicting_tags() {
+
+	// Push the first retracted fork block, with the ready tx.
+	let header01a = api.push_block_with_parent(header0.hash(), vec![xt0.clone()], true);
+	api.set_nonce(header01a.hash(), Alice.into(), 200);
+	let event = new_best_block_event(&pool, None, header01a.hash());
+	block_on(pool.maintain(event));
+	assert_pool_status!(header01a.hash(), &pool, 2, 0);
+	assert_eq!(api.validation_requests().len(), 5);
+
+	// Submit a second ready tx.
+	let _ = block_on(pool.submit_and_watch(invalid_hash(), SOURCE, xt3.clone())).unwrap();
+	let _ = block_on(pool.submit_and_watch(invalid_hash(), SOURCE, xt4.clone())).unwrap();
+	let _ = block_on(pool.submit_and_watch(invalid_hash(), SOURCE, xt5.clone())).unwrap();
+	let _ = block_on(pool.submit_and_watch(invalid_hash(), SOURCE, xt11.clone())).unwrap();
+	assert_pool_status!(header01a.hash(), &pool, 5, 1);
+	assert_eq!(api.validation_requests().len(), 9);
+
+	// Push the second retracted fork block, containing xt1.
+	let header02a = api.push_block_with_parent(header01a.hash(), vec![xt1.clone()], true);
+	api.set_nonce(header02a.hash(), Alice.into(), 201);
+	let event = new_best_block_event(&pool, None, header02a.hash());
+	block_on(pool.maintain(event));
+	assert_pool_status!(header02a.hash(), &pool, 4, 1);
+	assert_eq!(api.validation_requests().len(), 13);
+
+	// Submit another batch of future txs.
+	let _ = block_on(pool.submit_and_watch(invalid_hash(), SOURCE, xt9.clone())).unwrap();
+	let _ = block_on(pool.submit_and_watch(invalid_hash(), SOURCE, xt8.clone())).unwrap();
+	let _ = block_on(pool.submit_and_watch(invalid_hash(), SOURCE, xt7.clone())).unwrap();
+	assert_pool_status!(header02a.hash(), &pool, 4, 4);
+	assert_eq!(api.validation_requests().len(), 16);
+
+	// Push the third retracted fork block, containing xt1.
+	let header03a = api.push_block_with_parent(header02a.hash(), vec![xt2.clone()], true);
+	api.set_nonce(header03a.hash(), Alice.into(), 202);
+	let event = new_best_block_event(&pool, None, header03a.hash());
+	block_on(pool.maintain(event));
+	assert_pool_status!(header03a.hash(), &pool, 3, 4);
+	assert_eq!(api.validation_requests().len(), 19);
+
+	let header01b = api.push_block_with_parent(
+		header0.hash(),
+		vec![xt0.clone(), xt1.clone(), xt2.clone()],
+		true,
+	);
+	api.set_nonce(header01b.hash(), Alice.into(), 202);
+
+	let header02b = api.push_block_with_parent(
+		header01b.hash(),
+		vec![xt3.clone(), xt4.clone(), xt5.clone()],
+		true,
+	);
+	api.set_nonce(header02b.hash(), Alice.into(), 205);
+
+	let header03b = api.push_block_with_parent(
+		header02b.hash(),
+		vec![xt6.clone(), xt7.clone(), xt8.clone()],
+		true,
+	);
+	api.set_nonce(header03b.hash(), Alice.into(), 208);
+
+	let header04b = api.push_block_with_parent(
+		header03b.hash(),
+		vec![xt9.clone(), xt10.clone(), xt11.clone()],
+		true,
+	);
+	api.set_nonce(header04b.hash(), Alice.into(), 211);
+
+	// Notify a whole new fork.
+	let event = new_best_block_event(&pool, None, header04b.hash());
```

Comment:
I've just realized that pool is always computing the tree-route, but I think it would be good to provide the right tree-route in event:

```suggestion
	let event = new_best_block_event(&pool, Some(header03a.hash()), header04b.hash());
```

Could be also applied in the above events.